src/main/kotlin/com/serameracorp/pattern/PatternService.kt

```diff
@@ -0,0 +1,140 @@
+        if (fabricParam != null && fabricParam != "") {
+            val fabricId = fabricParam.toIntOrNull()
+            if (fabricId == null) {
+                throw IllegalArgumentException("Invalid fabricId: '$fabricParam'")
+            }
+            val fabricLength = fabricLengthParam?.toDoubleOrNull()
+            if (fabricLength == null) {
+                throw IllegalArgumentException("Invalid fabricLength: '$fabricLengthParam'")
+            }
+            patternRepository.createPatternFabricStatement.setInt(1, patternId)
+            patternRepository.createPatternFabricStatement.setInt(2, fabricId)
+            patternRepository.createPatternFabricStatement.setDouble(3, fabricLength)
+            patternRepository.createPatternFabricStatement.executeQuery()
+        }
+    }
+
+    fun findAllPatternsByName(searchParam: String?): List<Pattern> {
+        val resultSet = patternRepository.searchPatternStatement(searchParam).executeQuery()
+
+        return sequence {
+            while (resultSet.next()) {
+                yield(patternFromResultSet(resultSet))
+            }
+        }.toList()
+    }
+
+    fun findPatternWithDetails(patternId: Int): Pattern{
+        val pattern = findPatternById(patternId)
+        addPatternDetails(pattern)
+        return pattern
+    }
+
+    fun findPatternById(patternId: Int): Pattern {
+        patternRepository.patternByIdStatement.setInt(1, patternId)
+        val resultSet = patternRepository.patternByIdStatement.executeQuery()
+        if (resultSet.next()) {
+            // no fabric details yet
+            return patternDetailsFromResultSet(resultSet)
+        } else {
+            throw NotFoundException("The searched pattern was not found")
+        }
+    }
+
+    fun findProjectsByPatternId(patternId: Int): List<Project>{ //TODO move to project package?
+        patternRepository.projectByPatternStatement.setInt(1, patternId)
+        val projectResults = patternRepository.projectByPatternStatement.executeQuery()
+        return sequence {
+            while (projectResults.next()) {
+                yield(projectFromResultSet(projectResults))
+            }
+        }.toList()
+    }
+
+    fun addPatternDetails(pattern: Pattern){
+        pattern.patternFabric.addAll(findPatternFabrics(pattern.id))
+        pattern.clothingType.addAll(findPatternClothingTypes(pattern.id))
+    }
+
+    private fun findPatternClothingTypes(patternId: Int): List<ClothingType> {
+        patternRepository.patternClothingTypeByPatternIdStatement.setInt(1, patternId)
+        val clothingTypePatternResultSet = patternRepository.patternClothingTypeByPatternIdStatement.executeQuery()
+
+        val clothingTypes: MutableList<ClothingType> = mutableListOf()
+        while (clothingTypePatternResultSet.next()) {
+            clothingTypes.add(
+                ClothingType(clothingTypePatternResultSet.getString("clothing_type"))
+            )
+        }
+        return clothingTypes
```

Comment:
Could we use the `sequence` function here, as we did in `findProjectsByPatternId`?